File: sandbox/cerebusParsing/CerebusParse.m
classdef CerebusParse < coder.ExternalDependency
    %#codegen
    % Author: Dan O'Shea [ dan { at } djoshea.com ] (c) 2015
    
    methods (Static)
        function bName = getDescriptiveName(~)
            bName = 'CerebusParse';
        end
        
        function tf = isSupportedContext(ctx)
            if  ctx.isMatlabHostTarget()
                tf = true;
            else
                error('CerebusParse library not available for this target');
            end
        end
        
        function updateBuildInfo(buildInfo, ctx)
            % get path to this file
            stack = dbstack('-completenames');
            filePath = fileparts(stack(1).file);
            buildInfo.addIncludePaths(filePath);
            buildInfo.addIncludeFiles('cerebusParse.h');
            buildInfo.addSourcePaths(filePath);
            buildInfo.addSourceFiles('cerebusParse.c');
            
            % for nblib.h
            buildInfo.addIncludePaths(fullfile(matlabroot, 'toolbox\rtw\targets\xpc\target\build\xpcblocks\include'));
            buildInfo.addSourcePaths(fullfile(matlabroot, 'toolbox\rtw\targets\xpc\target\build\xpcblocks\'));
            
            % for cnblib.h
%             buildInfo.addIncludePaths(fullfile(matlabroot, 'toolbox\shared\xpc\canad\utils'));
%             buildInfo.addIncludePaths(fullfile(matlabroot, 'toolbox\shared\xpc\canad\utils\include'));
%             buildInfo.addSourcePaths(fullfile(matlabroot, 'toolbox\shared\xpc\canad\utils'));
%             buildInfo.addSourceFiles(fullfile(matlabroot, 'toolbox\shared\xpc\canad\utils\cnblib.c'));
            
%             [~, linkLibExt, ~, ~] = ctx.getStdLibInfo();
            
%             % Link files
%             linkPath = fullfile(matlabroot, 'toolbox\rtw\targets\xpc\target\build\xpcblocks\lib');
%             linkPriority = '';
%             linkPrecompiled = true;
%             linkLinkOnly = true;
%             group = '';
%             buildInfo.addLinkObjects('nblib_vc.lib', linkPath, ...
%                 linkPriority, linkPrecompiled, linkLinkOnly, group);
%             
%             buildInfo.addLinkObjects('xpcip_vc.lib', linkPath, ...
%                 linkPriority, linkPrecompiled, linkLinkOnly, group);
%             
%             buildInfo.addLinkObjects('etherlib_vc.lib', linkPath, ...
%                 linkPriority, linkPrecompiled, linkLinkOnly, group);
%             
%             buildInfo.addLinkObjects('ether8255x_vc.lib', linkPath, ...
%                 linkPriority, linkPrecompiled, linkLinkOnly, group);

%             buildInfo.addLinkObjects();
        end
        
        % use nbExtractPacketData when the data is stored
        % inside a network buffer. pp is the pointer to the cbPKT data that
        % gets passed to the other functions below. pBufferEnd is a pointer
        % to the end of the data in memory that is used by hasNext below.
        function [pp, pBufferEnd] = nbExtractPacketData(nbInput)
            coder.cinclude('cerebusParse.h');
            pp = coder.opaque('cbPKT_GENERIC*', 'NULL');
            pBufferEnd = coder.opaque('UINT8_T*', 'NULL');
            
            packetLen = uint32(0);
            pp = coder.ceval('cb_nbExtractPacketData', nbInput, coder.ref(packetLen));
                    
            % returns NULL on error
            if cast(pp, 'uint32') == 0
                error('CerebusParse: Could not extract packet data');
            end
            
            % grab pointer to end of data buffer
            pBufferEnd = coder.ceval('cb_getPointerToBufferEnd', pp, packetLen);
        end
        
        function nbFree(nbInput)
            tf = coder.nullcopy(uint8(0));
            tf = coder.ceval('cb_nbFree', nbInput);
            if tf
                error('CerebusParse: Error freeing packet data');
            end 
        end
        
        % use this function when you're manually passing in the data as a
        % uint8 vector. 
        % 
        % returns opaque cbPKT_GENERIC*, opaque UINT8_T*
        function [pp, pBufferEnd] = getPointersFromRawData(data)
            coder.cinclude('cerebusParse.h');
            assert(isa(data, 'uint8'));
            pp = coder.opaque('cbPKT_GENERIC*', 'NULL');
            pBufferEnd = coder.opaque('UINT8_T*', 'NULL');
            pp = coder.ceval('cb_getPointer', coder.ref(data));
            pBufferEnd = coder.ceval('cb_getPointerToBufferEnd', pp, uint32(numel(data)));
        end
        
        function [pp, pBufferEnd] = getPointersFromDataPtr(dataPtr, dataLen)
            coder.cinclude('cerebusParse.h');
            assert(isa(dataPtr, 'uint32'));
            pp = coder.opaque('cbPKT_GENERIC*', 'NULL');
            pBufferEnd = coder.opaque('UINT8_T*', 'NULL');
            pp = coder.ceval('(cbPKT_GENERIC*)', dataPtr); % do a cast operation
            pBufferEnd = coder.ceval('cb_getPointerToBufferEnd', pp, uint32(dataLen));
        end
        
        % returns logical as uint8
        function tf = hasNext(pp, pBufferEnd)
            coder.cinclude('cerebusParse.h');
            tf = coder.nullcopy(uint8(0));
            tf = coder.ceval('cb_hasNext', pp, pBufferEnd);
        end
        
        % returns opaque cbPKT_GENERIC*
        function pNext = getNext(pp)
            coder.cinclude('cerebusParse.h');
            pNext = coder.opaque('cbPKT_GENERIC*', 'NULL');
            pNext = coder.ceval('cb_getNext', pp);
        end
        
        % returns uint32
        function t = getTime(pp)
            coder.cinclude('cerebusParse.h');
            t = coder.nullcopy(uint32(0));
            t = coder.ceval('cb_getTime', pp);
        end
        
        % returns uint16
        function c = getChannel(pp)
            coder.cinclude('cerebusParse.h');
            c = coder.nullcopy(uint16(0));
            c = coder.ceval('cb_getChannel', pp);
        end
        
        % return logical as uint8
        function tf = isSpikePacket(pp)
            coder.cinclude('cerebusParse.h');
            tf = coder.nullcopy(uint8(0));
            tf = coder.ceval('cb_isSpikePacket', pp);
        end
        
        % returns logical as uint8
        function tf = isContinuousPacketForGroup(pp, group)
            coder.cinclude('cerebusParse.h');
            tf = coder.nullcopy(uint8(0));
            tf = coder.ceval('cb_isContinuousPacketForGroup', pp, uint8(group));
        end 
        
        % returns uint8
        function u = getSpikeUnit(pp)
            coder.cinclude('cerebusParse.h');
            u = coder.nullcopy(uint8(0));
            u = coder.ceval('cb_getSpikeUnit', pp);
        end
        
        % the buffer is returned from this function solely so that code
        % generation treats the buffer as mutable and doesn't make a copy
        % before passing it in
        function buffer = copySpikeWaveformToBufferColumn(pp, buffer, col)
            coder.cinclude('cerebusParse.h');
            assert(nargout > 0, 'Store the buffer output from this function so that it is properly written to');
            coder.ceval('cb_copySpikeWaveform', pp, coder.ref(buffer(1, col)), size(buffer, 1)); 
        end
            
        % the buffer is returned from this function solely so that code
        % generation treats the buffer as mutable and doesn't make a copy
        % before passing it in
        function buffer = copyContinuousSamplesToBufferColumn(pp, buffer, col)
            coder.cinclude('cerebusParse.h');
            assert(nargout > 0, 'Store the buffer output from this function so that it is properly written to');
            coder.ceval('cb_copyContinuousSamples', pp, coder.ref(buffer(1, col)), size(buffer, 1)); 
        end
    end
end

File: sandbox/cerebusParsing/cerebusParse.c
#include <stddef.h>
#include <stdlib.h>
#include "simstruc.h"

#ifndef MATLAB_MEX_FILE
#include <windows.h>
#include "xpctarget.h"
#endif

#ifdef MATLAB_MEX_FILE
#include "mex.h"
#endif

#include "nblib.h"
#include "utilities.c"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "cerebusParse.h"
#include "nblib.h"

#define MIN(a,b) (((a)<(b))?(a):(b))
#define MAX(a,b) (((a)>(b))?(a):(b))

/// xPC Networking utilities
// 
// // returns NULL on error
// cbPKT_GENERIC* cb_nbExtractPacketData(UINT32_T* nbInput, UINT32_T* pDataLen) {
//    // Get Buffer
//    xpcNBError error;
//    xpcNB* nb = (xpcNB*)nbInput;
//    INT32_T len;
//    cbPKT_GENERIC* pp;
// 
//    // Accept Buffer
//    if ( error = xpcNBAccept(nb) ) 
//    {
//       fprintf(stderr, "CerebusParse: Network Buffer Extract Accept Error %d", error);
//       return NULL;
//    }
// 
//    len = xpcNBBytes(nb);
//    if ( len < 0 ) len = 0;
// 
//    pp = (cbPKT_GENERIC*)xpcNBData(nb);
//    *pDataLen = len;
//    
//    return pp;
// }
// 
// // returns true on error
// bool cb_nbFree(UINT32_T* nbInput)
// {
//     xpcNB* nb = (xpcNB*)nbInput;
//     xpcNBError error;
//     // Free Buffer
//     if ( error = xpcNBFree(nb) ) {
//        fprintf(stderr, "CerebusParse: Network Buffer Extract Free Error %d", error);
//        return true;
//     }
//     return false;
// }

cbPKT_GENERIC* cb_getPointer(cbPKT_GENERIC* pp)
{
    return pp;
}

// return a pointer to the last byte in the buffer
UINT8_T* cb_getPointerToBufferEnd(cbPKT_GENERIC* pp, UINT32_T nBytes)
{
    return (UINT8_T*) (((UINT8_T*)pp) + (nBytes - 1));
}

// using the pointer to the last byte of the buffer, determine whether there
// is room for another cbPKT in the buffer. If not, return NULL. If there is
// return the pointer to the next packet
bool cb_hasNext(cbPKT_GENERIC*pp, UINT8_T* pBufferEnd)
{
    cbPKT_GENERIC* pNext = cb_getNext(pp);
    if((UINT8_T*)pNext >= pBufferEnd)
        return false;
    if(pNext->time == 0)
        return false;
    return true;
}   

cbPKT_GENERIC* cb_getNext(cbPKT_GENERIC*pp)
{
    return (cbPKT_GENERIC*)(((UINT8_T*)pp) + cbPKT_HEADER_SIZE + pp->dlen*4);
}

UINT32_T cb_getTime(cbPKT_GENERIC* pp)
{
    return pp->time;
}

UINT16_T cb_getChannel(cbPKT_GENERIC* pp)
{
    return pp->chid;
}

UINT8_T cb_getSpikeUnit(cbPKT_GENERIC* pp)
{
    return ((cbPKT_SPK*)pp)->unit;
}

bool cb_isSpikePacket(cbPKT_GENERIC* pp)
{
   return pp->chid > 0 && pp->chid <= cbNUM_ANALOG_CHANS;
}

bool cb_isContinuousPacketForGroup(cbPKT_GENERIC* pp, UINT8_T group)
{
   return pp->chid == 0 && pp->type == group;
}

UINT8_T cb_getContinuousGroup(cbPKT_GENERIC* pp) 
{
    return pp->type;
}

void cb_copySpikeWaveform(cbPKT_GENERIC* pp, INT16_T* buffer, int maxSamples)
{
    // ps->dlen indicates the number of 4 byte words in the 
    // fPattern (3*4), nPeak(2), nValley (2 bytes), and wave fields.
    // thus we want to copy (dlen-4)*4 bytes of data
    cbPKT_SPK* ps = (cbPKT_SPK*)pp;
    memcpy(buffer, ps->wave, MIN(maxSamples*sizeof(INT16_T), (ps->dlen-4) * 4));
}

void cb_copyContinuousSamples(cbPKT_GENERIC* pp, INT16_T* buffer, int maxChannels)
{
    // ps->dlen indicates the number of 4 byte words in the data
    cbPKT_GROUP* pg = (cbPKT_GROUP*)pp;
    memcpy(buffer, pg->data, MIN(maxChannels*sizeof(INT16_T), pg->dlen * 4));
}
